### ISSUE TITLE
Accumulating GPU memory during _embed_dl (?)

### ISSUE THREAD

**Issue by mhieronymus** [goergen95/seapig#76]
Looking at `seapig/scores/embed.py` I see the following issue
```
@classmethod
    def _embed_dl(
        self,
        model: torch.nn.Module,
        loader: DataLoader[torch.Tensor | dict[str, torch.Tensor]],
    ) -> torch.Tensor:
        """Extract embeddings by iterating over a DataLoader.

        This method ensures the model is in eval mode during embedding extraction
        to ensure consistent behavior regardless of the model's initial state.
        The model's original training state is restored after embedding extraction.
        """
        assert callable(model.embed)
        # Save the current training state and set model to eval mode
        was_training = model.training
        model.eval()

        pbar_desc = f"Embedding {len(loader)} batches"
        embs_ls = list()
        for batch in track(
            loader, total=len(loader), desc=pbar_desc, unit="batches"
        ):
            z = self._embed(X=batch, model=model)  # <- z is still on device
            embs_ls.append(z)
        embs = torch.cat(embs_ls, dim=0) # <- Accumulating memory on GPU until OoM

        # Restore the original training state
        if was_training:
            model.train()

        return embs
```
Unless it is expected that `model.embed()` also sends the tensor to CPU, this is leading to Out of Memory errors.  You create multiple `embs_ls` and a large `embs` on device. I recommend the following fix
```
embs_ls = []
for batch in loader:
    z = self._embed(X=batch, model=model)
    embs_ls.append(z.detach().cpu())
```
`detach()` removes the tensor from any computation graph (eval does not disable gradients), `cpu()` does nothing if already on cpu. 
Or change `_embed()` in `seapig/scores/embed.py` accordingly.

Sidenote:
Instead of `detach()`, you may also use `with torch.no_grad()`. Would be even better.

**Comment by goergen95** [comment]
Concerning the issue raised with accumulating gradients, since `_embed()` is called from within `_embed_dl()`  these tensors should not be connected to a computation graph because it is decorated with `@torch.inference_mode()` which the docs states is equivalent to [`no_grad`](https://docs.pytorch.org/docs/stable/generated/torch.autograd.grad_mode.inference_mode.html):

https://github.com/goergen95/seapig/blob/4a47fbdeca6f6c12ddb131a85aa51b96a5b40bc9/seapig/scores/embed.py#L119-L120

I do see that this could fill up the device's memory (this could also happen on CPU, though we can expect more memory available there). A fix could be to transfer to `z.cpu()` in `_embed()` though I am a little bit hesitant to push this just yet  because I would have to evaluate if this causes any device mismatches further down. Are you actually already running in this issue or did you just expect it to occur from reading the code?

**Comment by mhieronymus** [comment]
I did not test it on GPU yet. I just saw the code and some of my models use up 70 GB of VRAM easily with a batch size of 2 already. Unless this supports multi-GPU, already, thatmight hurt down the line. I just opened the issue for reference in case I run into that. Feel free to ignore me unless I ping you or provide a pull request. :)

**Comment by goergen95** [comment]
Yes, please ping me if you run into a wall here. My current thinking is as follows: embeddings are currently expected to be of size `(N,D)`, so even with large D the memory footprint is significantly less than the compared to the actual inputs. That is also the reason why `fit()` either expects pre-computed reference embedding, or collects all embeddings on the fly before populating the index. Depending on N, however, I see that the current approach can run into the memory issue you expect here. If we hit that wall, I would opt to implement a true `batched` approach instead, which would mean to populate the index batch-wise. 

Current `_build_index()`  implementation for reference:
https://github.com/goergen95/seapig/blob/4a47fbdeca6f6c12ddb131a85aa51b96a5b40bc9/seapig/scores/knn.py#L217-L243

**Comment by mhieronymus** [comment]
Well, depending on the work we do, we have a couple of TB for training. Even with a lower embedded space, this explodes. And if anyone else at DWD gets interested in this, we can easily go into the PBs of the training data.
But I'll try with my smaller datasets first and give feedback.

**Comment by goergen95** [comment]
Kind of outdated, but we now move to CPU when accumulating tensors in the extractor class, thus closing here.

https://github.com/goergen95/seapig/blob/2af2ad7a1c7799e5be5b68142b3ab7eced62f934/seapig/scores/extractor.py#L226
